devlib/utils/ssh.py

```diff
@@ -170,51 +305,422 @@ def name(self):
+
+        self.client = self._make_client()
+        atexit.register(self.close)
+
+    def _make_client(self):
+        if self.strict_host_check:
+            policy = RejectPolicy
+        else:
+            policy = AutoAddPolicy
+
+        with _handle_paramiko_exceptions():
+            client = SSHClient()
+            client.load_system_host_keys()
+            client.set_missing_host_key_policy(policy)
+            client.connect(
+                hostname=self.host,
+                port=self.port,
+                username=self.username,
+                password=self.password,
+                key_filename=self.keyfile,
+                timeout=self.timeout,
+            )
+
+            return client
+
+    def _make_channel(self):
+        with _handle_paramiko_exceptions():
+            transport = self.client.get_transport()
+            channel = transport.open_session()
+            return channel
+
+    def _get_sftp(self, timeout):
+        sftp = self.client.open_sftp()
+        sftp.get_channel().settimeout(timeout)
+        return sftp
+
+    @classmethod
+    def _push_file(cls, sftp, src, dst):
+        sftp.put(src, dst)
+
+    @classmethod
+    def _push_folder(cls, sftp, src, dst):
+        # This might fail if the folder already exists
+        with contextlib.suppress(IOError):
+            sftp.mkdir(dst)
+        for entry in os.scandir(src):
+            name = entry.name
+            src_path = os.path.join(src, name)
+            dst_path = os.path.join(dst, name)
+            if entry.is_dir():
+                push = cls._push_folder
+            else:
+                push = cls._push_file
+
+            push(sftp, src_path, dst_path)
+
+    @classmethod
+    def _push_path(cls, sftp, src, dst):
+        push = cls._push_folder if os.path.isdir(src) else cls._push_file
+        push(sftp, src, dst)
+
+    @classmethod
+    def _pull_file(cls, sftp, src, dst):
+        with contextlib.suppress(FileNotFoundError):
+            os.remove(dst)
+        sftp.get(src, dst)
+
+    @classmethod
+    def _pull_folder(cls, sftp, src, dst):
+        with contextlib.suppress(FileNotFoundError):
+            try:
+                shutil.rmtree(dst)
+            except OSError:
+                os.remove(dst)
+
+        os.makedirs(dst)
+        for fileattr in sftp.listdir_attr(src):
+            filename = fileattr.filename
+            src_path = os.path.join(src, filename)
+            dst_path = os.path.join(dst, filename)
+            if stat.S_ISDIR(fileattr.st_mode):
+                pull = cls._pull_folder
+            else:
+                pull = cls._pull_file
+
+            pull(sftp, src_path, dst_path)
+
+    @classmethod
+    def _pull_path(cls, sftp, src, dst):
+        try:
+            cls._pull_file(sftp, src, dst)
+        except IOError:
+            # Maybe that was a directory, so retry as such
+            cls._pull_folder(sftp, src, dst)
```

Comment:
Rewording to make sure I understand correctly:
* _pull_file or _pull_folder will be selected based on what there is to pull
* if a file is pulled in a folder, the src basename will be appended to the dst folder

Reply:
- Agreed, the existing functionality seems to identify this correctly.
- That at least is what makes sense to me to mimic existing behaviour of the other connections. For example within WA we frequently just want to pull file X into our output directory rather than explicitly creating the full file paths for both source and destination.

Reply:
Should pushing an src folder into a dst folder do:
1) Push the content of src into dst
2) "move" src into dst, which means creating dst/src.basename

The 1st behavior seems more sensible and make pushing folder idempotent, which I think is very desirable. The 2nd behavior corresponds to the mv command.

Reply:
Good question. The first option might seem useful however just looking at the other implementations and it looks like the 2nd option is more consistent with the `AdbConnection` and the existing `SshConnection` behaviour so unsure it would make sense for this version to operate differently?

Reply:
Looks like adb actually behaves like 1):
```
$ adb push . /mnt/sdcard/
push: ./t2.txt -> /mnt/sdcard/t2.txt
push: ./t3.txt -> /mnt/sdcard/t3.txt
push: ./t1.txt -> /mnt/sdcard/t1.txt
3 files pushed. 0 files skipped.
```
https://stackoverflow.com/questions/13145872/android-recursive-copy-with-adb-push